# Lesson 2: Solutions
These exercises are to help you check your understanding during class. They will not be collected.

## Part 1: JupyterHub

Your Name

### Exercise 2 - Create a cell and compute a sum

* Insert a cell
    + Select this cel by clicking it once.
    + Use the menue option Insert > Insert Cell Below to insert a cell below this one.
* Compute a sum
    + Double click on the new cell to edit it
    + Type 12 + 7 in the new cell
    + Press SHIFT + Enter to exicute.


In [1]:
12+7

[1] 19

### Exercise 3 - Change cell type  and compute a percentage

* Change the markdown cell to a code cell
    + Select the cell by clicking on it once
    + use the menu option Cell > Cell Type > Code
* Suppose that out of 60 students, 25 are living in Manhattan, 11 in Brooklyn, 9 in Queens, 9 in the Bronx, 3 in Staten Island, 2 in New Jersey, and 1 in Westchester. Use the cell below to compute the percent of of students who live in Brooklyn.

In [6]:
# Percentage of students in Brooklyn 5 +2
5+2
11/60*100

[1] 7

[1] 18.33333

18.33% of the students live in Brooklyn.

## Part 2: Compute

### Exercise 4: Arithmetic
Verify the following equation by using the code cell below to comput the value of the left hand side.
$$\frac{3^2 + 4^2}{5} = 5$$

In [41]:
(3^2+4^2)/5

[1] 5

### Exercise 5: Square root
Verify the following equation by using the code cell below to comput the value of the left hand side.
$$\sqrt{\lvert 5^2 - 13^2\rvert} = 12$$

In [42]:
sqrt(abs(5^2-13^2))

[1] 12

### Exercise 6: Logarightm
Verify the following equation by using the code cell below to comput the value of the left hand side.
$$\ln(e^\pi) \approx 3.14159265358979$$

Inside the $\ln$ we have $e^\pi$

In [7]:
exp(pi)

[1] 23.14069

In [8]:
log(23.1406926327793)

[1] 3.141593

In [40]:
log(exp(pi))

[1] 3.141593

## Part 3: Vectors

### Exercise 7: Boroughs
Create a vect0r that contains the names of the five boroughs of New York city:
- Brooklyn
- Bronx
- Manhattan
- Queens
- Staten Island

In [43]:
c('Brooklyn', 'Bronx', 'Manhattan', 'Queens', 'Staten Island')

[1] "Brooklyn"      "Bronx"         "Manhattan"     "Queens"       
[5] "Staten Island"

### Exercise 8: Odd numbers
Definition: An *even* number is any number that can be expressed as $2n$ where $n$ is an integer. 

Definition: An *odd* number is any number that can be expressed as $2n+1$ where $n$ is an integer. 

In Example 6, we used arithmeic operations on lists to created a list of the even numbers beteen 1 and 20. Use a similar strategy to create a list of the odd numbers between 1 and 20.

In [69]:
(1:10)*2 - 1

[1]  1  3  5  7  9 11 13 15 17 19

In [70]:
(0:9)*2 + 1

[1]  1  3  5  7  9 11 13 15 17 19

In [9]:
(1:10)*2 +1

[1]  3  5  7  9 11 13 15 17 19 21

In [10]:
c(1,2,3)*2 +1

[1] 3 5 7

### Exercise 9: Combining vectors

The cell below contains two vectors. The first vector contains the predicted high temperature in degrees Fahrenheit in NYC for the next five days. The second vector contains the predicted low temperatures. 

Use vector operations to find the average of the high and the low temerature for each day. 

In [ ]:
# The pound sybol is used for comments. 
# R ignores everything in the line after the # symbol.

# highs 
c(35, 39, 42, 36, 35)

# lows
c(27, 29, 37, 38, 20)

Average $$\frac{h + \ell}{2}$$

In [74]:
(c(35, 39, 42, 36, 35) + c(27, 29, 37, 38, 20))/2

[1] 31.0 34.0 39.5 37.0 27.5

## Examples

### Example 1: Arithmetic
R uses the standard order of operations.

$$1 + 4 \cdot 3 ^ 2 - \frac{5}{2} = 34.5$$

In [8]:
1 + 4 * 3 ^ 2 - 5/2

[1] 34.5

$$\frac{1 + (4 \cdot 3) ^ 2 - 10}{2}=70$$

In [7]:
(1 + (4*3)^2 - 5)/2

[1] 70

### Example 2: Mathematical constants and functions

- ractio of the circumfance and diameter of circle: `pi`, $\pi \approx 3.14159265358979$

In [11]:
pi

[1] 3.141593

- base of the natural logarithm: `exp(1)`, $e \approx 2.71828182845905$

In [14]:
exp(1)

[1] 2.718282

- absolute value:  `abs()` eg. $\lvert -5 \rvert = 5$

In [9]:
abs(-5)

[1] 5

- square root: `sqrt()` eg. $\sqrt{9} = 3$

In [10]:
sqrt(9)

[1] 3

- natural logarithm: `log()` eg. $\ln(2.71) \approx 1$

In [18]:
log(2.71)

[1] 0.9969486

- exponential function: `exp()` eg. $e^3 \approx 20.0855369231877$

In [19]:
exp(3)

[1] 20.08554

### Example 3: Data types
The `typeof()` returns the data type of the R object that is inputed into it. 

- **character** refers to letters and symbols in quote marks
- **double** is a decimal number, note $5 = 5.0$ can be thought of as a decimal
- **integer** is a whole number eg. $\dots , -2,-1,0,1,2, \dots $

In [23]:
typeof('smile')

[1] "character"

In [49]:
typeof(5)

[1] "double"

### Example 4: Vectors
Vectors are ordered collections of entries where **all** entries have the same data type. The R function `c()` is used to construct vectors. The `typof()` function returns the type of the elements in the vector.

In [36]:
c('a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z')

[1] "a" "b" "c" "d" "e" "f" "g" "h" "i" "j" "k" "l" "m" "n" "o" "p" "q" "r" "s"
[20] "t" "u" "v" "w" "x" "y" "z"

In [44]:
c(2,3,5,7,11,13,17,19)

[1]  2  3  5  7 11 13 17 19

In [47]:
typeof(c('a','b','c'))

[1] "character"

Note: R will coerce the elements of a vector into the same type. In the example below, the double 1 is converted into a chacter '1'.

In [45]:
c('a', 1)

[1] "a" "1"

### Example 5: Vectors of numbers
If $a$ and $b$ are integers such that $a\le b$, then `a:b` wich create the vector of integers starting with $a$ and ending with $b$. 

In [51]:
1:10

[1]  1  2  3  4  5  6  7  8  9 10

In [53]:
-5:2

[1] -5 -4 -3 -2 -1  0  1  2

### Example 6: Math with one vector
Arithmetic operations are applied to each element of a vector.

In [60]:
1:10 + 2

[1]  3  4  5  6  7  8  9 10 11 12

In [59]:
(1:10)*2

[1]  2  4  6  8 10 12 14 16 18 20

In [61]:
(1:10)^2

[1]   1   4   9  16  25  36  49  64  81 100

### Example 7: Math with multiple vectors
Arithmetic operations with multiple vectors are done entry-by-entry.

In [75]:
c(2,4,6,8) + c(-1,0,1,2)

[1]  1  4  7 10

In [73]:
c(2 + (-1), 4 + 0, 6 + 1, 8 + 2) # sum of vectors done manualy

[1]  1  4  7 10